# Business Problem
Currently, if anyone wants to sell their car either they have to take their car to a respective company workshop or have to make an appointment for the company to get an estimate of the price. This process involves a lot of time and resources. Our Objective is to make a model for third-party companies that will make an estimate of the Price of the customer’s car directly from their online portal. This will save customers time and help the company to reduce its cost and also streamline the process of selling used cars. 
## Outlines 
#### 1. Data cleaning 
#### 2. feature engineering 
#### 3. Getting more insights 
#### 4. Data Pre-processing
- Splitting  
- fearure transformation 
#### 5. modeling 
#### 6. evalution 

## Import Packages

In [1]:
# Import the NumPy library and alias it as np
import numpy as np

# Import the Pandas library and alias it as pd
import pandas as pd

# Import the LinearRegression class from the scikit-learn library
from sklearn.linear_model import LinearRegression

# Import the StandardScaler and PolynomialFeatures classes from scikit-learn
from sklearn.preprocessing import StandardScaler, PolynomialFeatures

# Import the train_test_split function from scikit-learn for splitting the data
from sklearn.model_selection import train_test_split

# Import the plotly Express and FigureFactory modules for interactive plotting
import plotly.express as px
import plotly.figure_factory as ff


In [2]:
df = pd.read_csv("train.csv")
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  
0        NaN   1.75  
1        NaN  12.50  
2  8.61 Lakh   4.50  
3        NaN   6.00  
4        NaN  17.74

# 1. Data cleaning 


### 1.1 Investigation  

In [3]:
# Display information about the DataFrame, including data types and memory usage
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6019 entries, 0 to 6018
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               6019 non-null   object 
 1   Location           6019 non-null   object 
 2   Year               6019 non-null   int64  
 3   Kilometers_Driven  6019 non-null   int64  
 4   Fuel_Type          6019 non-null   object 
 5   Transmission       6019 non-null   object 
 6   Owner_Type         6019 non-null   object 
 7   Mileage            6017 non-null   object 
 8   Engine             5983 non-null   object 
 9   Power              5983 non-null   object 
 10  Seats              5977 non-null   float64
 11  New_Price          824 non-null    object 
 12  Price              6019 non-null   float64
dtypes: float64(2), int64(2), object(9)
memory usage: 611.4+ KB


In [4]:
# Check for missing values in the DataFrame and display the sum of null values for each column
df.isnull().sum()

Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 36
Power                  36
Seats                  42
New_Price            5195
Price                   0
dtype: int64

In [5]:
# Generate descriptive statistics of the DataFrame, including count, mean, min, and more
df.describe()

Year  Kilometers_Driven        Seats        Price
count  6019.000000       6.019000e+03  5977.000000  6019.000000
mean   2013.358199       5.873838e+04     5.278735     9.479468
std       3.269742       9.126884e+04     0.808840    11.187917
min    1998.000000       1.710000e+02     0.000000     0.440000
25%    2011.000000       3.400000e+04     5.000000     3.500000
50%    2014.000000       5.300000e+04     5.000000     5.640000
75%    2016.000000       7.300000e+04     5.000000     9.950000
max    2019.000000       6.500000e+06    10.000000   160.000000

In [6]:
# Check for duplicated rows in the DataFrame and display the total number of duplicates
df.duplicated().sum()

0

### Notes

- **Numerical Values Mixed with Text:**
  - The columns (Mileage, Engine, Power) contain a mix of numerical values and text.
  - In the Mileage column, there are two units present: kmpl and km/kg.

- **New_Price Column:**
  - A significant portion of the New_Price column consists of null values.
  - Two possible solutions:
    - Remove the column entirely.
    - Explore options to impute or fill in some data.

- **Null Values in Other Columns:**
  - Null values are present in other columns.

- **Seats Column:**
  - The Seats column contains some values that are zero.

- **Duplicated Rows:**
  - Duplicated rows have been identified in the dataset.

- **Power Column:**
  - The Power column includes values like "null bhp."### Notes

- **Numerical Values Mixed with Text:**
  - The columns (Mileage, Engine, Power) contain a mix of numerical values and text.
  - In the Mileage column, there are two units present: kmpl and km/kg.

- **New_Price Column:**
  - A significant portion of the New_Price column consists of null values.
  - Two possible solutions:
    - Remove the column entirely.
    - Explore options to impute or fill in some data.

- **Null Values in Other Columns:**
  - Null values are present in other columns.

- **Seats Column:**
  - The Seats column contains some values that are zero.

- **Duplicated Rows:**
  - Duplicated rows have been identified in the dataset.

- **Power Column:**
  - The Power column includes values like "null bhp."

###  1.2 Working with data issues 

> Drop "New_Price" column because most of them is null 

In [7]:
# Drop the 'New_Price' column
df.drop('New_Price',axis=1, inplace=True)

> Why did I choose to remove Null values and not to replace it ? 
- That's simply because this columns need to be fixed and we cannot calculate any value before we fix it. 
- another solution is to make a place holder and replace it after fixing data, (Your desicion id depende on the time) 
- and this column have  5195 Null-value which is big number

In [8]:
#Drop rows with missing values (NaN) from the DataFrame
df.dropna(inplace=True)

In [9]:
# Display the count of unique values in the 'Power' column
df.Power.value_counts()

Power
74 bhp        235
98.6 bhp      131
73.9 bhp      125
140 bhp       123
78.9 bhp      111
             ... 
201 bhp         1
199.3 bhp       1
95 bhp          1
152.88 bhp      1
181.04 bhp      1
Name: count, Length: 371, dtype: int64

In [10]:
# Display the unique values in the 'Power' column
df.Power.unique()

array(['58.16 bhp', '126.2 bhp', '88.7 bhp', '88.76 bhp', '140.8 bhp',
       '55.2 bhp', '63.1 bhp', '171.5 bhp', '103.6 bhp', '74 bhp',
       '103.25 bhp', '116.3 bhp', '187.7 bhp', '115 bhp', '175.56 bhp',
       '98.6 bhp', '83.8 bhp', '167.62 bhp', '190 bhp', '88.5 bhp',
       '177.01 bhp', '80 bhp', '67.1 bhp', '102 bhp', '108.45 bhp',
       '138.1 bhp', '184 bhp', '179.5 bhp', '103.5 bhp', '64 bhp',
       '82 bhp', '254.8 bhp', '73.9 bhp', '46.3 bhp', '37.5 bhp',
       '77 bhp', '82.9 bhp', '149.92 bhp', '138.03 bhp', '112.2 bhp',
       '163.7 bhp', '71 bhp', '105 bhp', '174.33 bhp', '75 bhp',
       '103.2 bhp', '53.3 bhp', '78.9 bhp', '147.6 bhp', '147.8 bhp',
       '68 bhp', '186 bhp', '170 bhp', '69 bhp', '140 bhp', '78 bhp',
       '194 bhp', '500 bhp', '108.5 bhp', '86.8 bhp', '187.74 bhp',
       'null bhp', '132 bhp', '86.7 bhp', '73.94 bhp', '117.3 bhp',
       '218 bhp', '168.5 bhp', '89.84 bhp', '110 bhp', '90 bhp',
       '82.85 bhp', '67 bhp', '241.4 bhp', '3

In [11]:
# Filter the DataFrame to include only rows where 'Power' is 'null bhp'
Filter_df = df[df['Power'] == 'null bhp']
Filter_df

Name   Location  Year  \
76                Ford Fiesta 1.4 SXi TDCi     Jaipur  2008   
79                  Hyundai Santro Xing XL  Hyderabad  2005   
89                  Hyundai Santro Xing XO  Hyderabad  2007   
120   Hyundai Santro Xing XL eRLX Euro III     Mumbai  2005   
143    Hyundai Santro Xing XO eRLX Euro II      Kochi  2008   
...                                    ...        ...   ...   
5861                Hyundai Santro Xing XO    Chennai  2007   
5873   Hyundai Santro Xing XO eRLX Euro II       Pune  2006   
5925           Skoda Laura Classic 1.8 TSI       Pune  2010   
5943               Mahindra Jeep MM 540 DP    Chennai  2002   
5985                       Toyota Etios GD      Delhi  2013   

      Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage  \
76               111111    Diesel       Manual      First   17.8 kmpl   
79                87591    Petrol       Manual      First    0.0 kmpl   
89                73745    Petrol       Manual      First   17.0 kmpl   
120              102000    Petrol       Manual     Second   17.0 kmpl   
143               80759    Petrol       Manual      Third   17.0 kmpl   
...                 ...       ...          ...        ...         ...   
5861              79000    Petrol       Manual      First   17.0 kmpl   
5873              47200    Petrol       Manual     Second   17.0 kmpl   
5925              85000    Petrol       Manual      First   17.5 kmpl   
5943              75000    Diesel       Manual      First    0.0 kmpl   
5985              70000    Diesel       Manual      First  23.59 kmpl   

       Engine     Power  Seats  Price  
76    1399 CC  null bhp    5.0   2.00  
79    1086 CC  null bhp    5.0   1.30  
89    1086 CC  null bhp    5.0   2.10  
120   1086 CC  null bhp    5.0   0.85  
143   1086 CC  null bhp    5.0   1.67  
...       ...       ...    ...    ...  
5861  1086 CC  null bhp    5.0   1.85  
5873  1086 CC  null bhp    5.0   1.20  
5925  1798 CC  null bhp    5.0   2.85  
5943  2112 CC  null bhp    6.0   1.70  
5985  1364 CC  null bhp    5.0   3.88  

[103 rows x 12 columns]

In [12]:
# Get the indices of rows where 'Power' is 'null bhp'
indx = Filter_df.index
indx

Index([  76,   79,   89,  120,  143,  227,  245,  262,  307,  308,
       ...
       5529, 5533, 5647, 5755, 5759, 5861, 5873, 5925, 5943, 5985],
      dtype='int64', length=103)

In [13]:
# Drop rows with 'null bhp' in the 'Power' column,
df.drop(indx,axis=0,inplace=True)

- (KMPL) is refered to Kilometers Per Litre
- (km/kg) is refered to kilometers Per kilogram
> 1 liter  = 1 kilogram 

In [14]:
dirty_cols = ['Mileage', 'Engine', 'Power']
df[dirty_cols]

Mileage   Engine      Power
0     26.6 km/kg   998 CC  58.16 bhp
1     19.67 kmpl  1582 CC  126.2 bhp
2      18.2 kmpl  1199 CC   88.7 bhp
3     20.77 kmpl  1248 CC  88.76 bhp
4      15.2 kmpl  1968 CC  140.8 bhp
...          ...      ...        ...
6014   28.4 kmpl  1248 CC     74 bhp
6015   24.4 kmpl  1120 CC     71 bhp
6016   14.0 kmpl  2498 CC    112 bhp
6017   18.9 kmpl   998 CC   67.1 bhp
6018  25.44 kmpl   936 CC   57.6 bhp

[5872 rows x 3 columns]

In [15]:
# For each column in the list, convert the values to float by extracting the numeric part
for col in dirty_cols:
    df[col] = df[col].apply(lambda x : float(x.split()[0]))
df[dirty_cols]

Mileage  Engine   Power
0       26.60   998.0   58.16
1       19.67  1582.0  126.20
2       18.20  1199.0   88.70
3       20.77  1248.0   88.76
4       15.20  1968.0  140.80
...       ...     ...     ...
6014    28.40  1248.0   74.00
6015    24.40  1120.0   71.00
6016    14.00  2498.0  112.00
6017    18.90   998.0   67.10
6018    25.44   936.0   57.60

[5872 rows x 3 columns]

#### values health checking 

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5872 entries, 0 to 6018
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5872 non-null   object 
 1   Location           5872 non-null   object 
 2   Year               5872 non-null   int64  
 3   Kilometers_Driven  5872 non-null   int64  
 4   Fuel_Type          5872 non-null   object 
 5   Transmission       5872 non-null   object 
 6   Owner_Type         5872 non-null   object 
 7   Mileage            5872 non-null   float64
 8   Engine             5872 non-null   float64
 9   Power              5872 non-null   float64
 10  Seats              5872 non-null   float64
 11  Price              5872 non-null   float64
dtypes: float64(5), int64(2), object(5)
memory usage: 596.4+ KB


___________________________________

# 2. Feature Engineering

In [17]:
# check unique names
count_of_uniqueName = df.Name.nunique()
count_of_uniqueName

1811

In [18]:
# Print a message indicating the percentage of unique values in the 'Name' column of the DataFrame
print('Percentage of uniques', round(count_of_uniqueName/df.shape[0]*100),'%')

Percentage of uniques 31 %


In [19]:
# Create a new column 'Brand'
df['Brand']= df['Name'].apply(lambda x: x.split()[0])

In [20]:
df.Brand.nunique()


30

In [21]:
# let's change name column with just first 2 words
df['Name']= df.Name.apply(lambda x: ' '.join(x.split()[:2]))


In [22]:
count_of_uniqueName = df.Name.nunique()
# Print a message indicating the percentage of unique values in the 'Name' column of the DataFrame
print('Percentage of uniques', round(count_of_uniqueName/df.shape[0]*100),'%')

Percentage of uniques 4 %


In [23]:
df.head()

Name    Location  Year  Kilometers_Driven Fuel_Type Transmission  \
0   Maruti Wagon      Mumbai  2010              72000       CNG       Manual   
1  Hyundai Creta        Pune  2015              41000    Diesel       Manual   
2     Honda Jazz     Chennai  2011              46000    Petrol       Manual   
3  Maruti Ertiga     Chennai  2012              87000    Diesel       Manual   
4        Audi A4  Coimbatore  2013              40670    Diesel    Automatic   

  Owner_Type  Mileage  Engine   Power  Seats  Price    Brand  
0      First    26.60   998.0   58.16    5.0   1.75   Maruti  
1      First    19.67  1582.0  126.20    5.0  12.50  Hyundai  
2      First    18.20  1199.0   88.70    5.0   4.50    Honda  
3      First    20.77  1248.0   88.76    7.0   6.00   Maruti  
4     Second    15.20  1968.0  140.80    5.0  17.74     Audi

________________________

# 3. Data Analysis 

In [24]:
df.columns

Index(['Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type',
       'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats',
       'Price', 'Brand'],
      dtype='object')

### value_counts for some columns

In [25]:
from IPython.display import display, Markdown

# Display value counts for 'Brand'
display(Markdown("### Brand Value Counts"))
display(df['Brand'].value_counts())

# Display value counts for 'Name'
display(Markdown("### Name Value Counts"))
display(df['Name'].value_counts())

# Display value counts for 'Location'
display(Markdown("### Location Value Counts"))
display(df['Location'].value_counts())

# Display value counts for 'Fuel_Type'
display(Markdown("### Fuel_Type Value Counts"))
display(df['Fuel_Type'].value_counts())


### Brand Value Counts

Brand
Maruti           1175
Hyundai          1058
Honda             600
Toyota            394
Mercedes-Benz     316
Volkswagen        314
Ford              294
Mahindra          268
BMW               262
Audi              235
Tata              183
Skoda             172
Renault           145
Chevrolet         120
Nissan             89
Land               57
Jaguar             40
Mitsubishi         27
Mini               26
Fiat               23
Volvo              21
Porsche            16
Jeep               15
Datsun             13
Force               3
ISUZU               2
Ambassador          1
Isuzu               1
Bentley             1
Lamborghini         1
Name: count, dtype: int64

### Name Value Counts

Name
Maruti Swift       327
Honda City         265
Hyundai i20        247
Hyundai Verna      170
Toyota Innova      164
                  ... 
Mahindra XUV300      1
Porsche Boxster      1
Nissan Teana         1
Mini Countryman      1
Jaguar F             1
Name: count, Length: 207, dtype: int64

### Location Value Counts

Location
Mumbai        775
Hyderabad     718
Kochi         645
Coimbatore    629
Pune          594
Delhi         545
Kolkata       521
Chennai       476
Jaipur        402
Bangalore     347
Ahmedabad     220
Name: count, dtype: int64

### Fuel_Type Value Counts

Fuel_Type
Diesel    3152
Petrol    2655
CNG         55
LPG         10
Name: count, dtype: int64

### Q1 : What is the average of price for each Brand?

In [26]:
brand_price_mean  = df.groupby("Brand")["Price"].mean().reset_index().sort_values(by="Price",ascending=False)
brand_price_mean

Brand       Price
15    Lamborghini  120.000000
3         Bentley   59.000000
23        Porsche   49.204375
16           Land   39.861053
13         Jaguar   37.632250
19  Mercedes-Benz   26.917848
20           Mini   26.896923
1            Audi   25.569787
2             BMW   25.439618
12          Isuzu   20.000000
29          Volvo   18.802857
14           Jeep   18.718667
11          ISUZU   12.045000
27         Toyota   11.907970
21     Mitsubishi   11.058889
7           Force    9.333333
17       Mahindra    8.058955
25          Skoda    7.586453
8            Ford    6.957721
24        Renault    5.799034
10        Hyundai    5.513715
9           Honda    5.424167
28     Volkswagen    5.306815
22         Nissan    4.784719
18         Maruti    4.554868
26           Tata    3.605574
6            Fiat    3.570435
4       Chevrolet    3.057333
5          Datsun    3.049231
0      Ambassador    1.350000

In [27]:
px.histogram(brand_price_mean,x="Brand",y="Price",text_auto=True,color="Brand",title="Average of price for each Brand")

### Q2 : Is  the Transmission of car affects the price of cars?

In [28]:
transmission_price_mean = df.groupby("Transmission")["Price"].mean().reset_index().sort_values(by="Price", ascending=False)
transmission_price_mean

Transmission      Price
0    Automatic  19.914918
1       Manual   5.395448

In [29]:
fig = px.bar(
    transmission_price_mean,
    x="Transmission",
    y="Price",
    title="Average Price by Transmission Type",
    labels={"Transmission": "Transmission Type", "Price": "Average Price"},
)

# Show the plot
fig.show()


### Q3 : what is the Most types offered for sale?

In [30]:
df["count"] = 1

In [31]:
brand_count_sum = df.groupby("Brand")["count"].sum().reset_index().sort_values(by="count",ascending=False)
""" 
Other way to find the most popular brand? But we will use count column in other case.
df["Brand"].value_counts().reset_index().sort_values(by="count", ascending=False).head(1)
""" 
brand_count_sum.head(1)


Brand  count
18  Maruti   1175

In [32]:
fig = px.bar(brand_count_sum,
             x="Brand",
             y="count",
             title="Brand with the Highest Count"
)

# Show the plot
fig.show()

### Q4 : is her any  correlation between the price of Car and the Power of car?

In [33]:
correlation_coefficient = df['Price'].corr(df['Power'])

# Print the correlation coefficient
print(f"Correlation coefficient between Price and Power: {correlation_coefficient:.2f}")

Correlation coefficient between Price and Power: 0.77


In [34]:
fig = px.scatter(df, x='Power', y='Price', trendline="ols", opacity=0.9,title="Scatter Plot  Price vs Power")
# Show the plot
fig.show()

### Q5 :Average Price for each Fuel_Type

In [35]:
fuel_type_price_mean = df.groupby("Fuel_Type")["Price"].mean().reset_index().sort_values(by="Price",ascending=False)
fuel_type_price_mean

Fuel_Type      Price
1    Diesel  12.958496
3    Petrol   5.773416
0       CNG   3.558909
2       LPG   2.487000

In [36]:
fig = px.bar(fuel_type_price_mean,
             x="Fuel_Type",
             y="Price",
             title="Average Price for Each Fuel Type")

# Show the plot
fig.show()

### Q6 :The number of cars for sale everywhere

In [37]:
location_car_count  = df.groupby("Location")["count"].sum().reset_index().sort_values(by="count",ascending=False)
location_car_count

Location  count
9       Mumbai    775
5    Hyderabad    718
7        Kochi    645
3   Coimbatore    629
10        Pune    594
4        Delhi    545
8      Kolkata    521
2      Chennai    476
6       Jaipur    402
1    Bangalore    347
0    Ahmedabad    220

In [38]:
fig = px.bar(location_car_count,
             x="Location",
             y="count",
             title="Number of Cars for Sale in Each Location")

# Show the plot
fig.show()

### Q7 :The most popular type of car offered for sale every year

In [39]:
popular_cars_by_year  =df.groupby(["Year","Brand"])["count"].sum().reset_index().sort_values(by="count",ascending=False)
popular_cars_by_year.head(10)

Year    Brand  count
232  2016   Maruti    179
228  2016  Hyundai    156
209  2015   Maruti    150
187  2014   Maruti    145
205  2015  Hyundai    142
256  2017   Maruti    132
249  2017  Hyundai    123
183  2014  Hyundai    122
182  2014    Honda    105
139  2012  Hyundai    104

In [40]:
px.histogram(popular_cars_by_year,x="Brand",y="count",color="Year",text_auto=True,title="The most popular type of car offered for sale every year")

### Q8 : The Transmission of car has changed over the years

In [41]:
df.groupby(["Year","Transmission"])["count"].sum().reset_index().sort_values(by="count",ascending=False)

Year Transmission  count
29  2014       Manual    570
31  2015       Manual    519
33  2016       Manual    509
27  2013       Manual    460
25  2012       Manual    420
35  2017       Manual    388
23  2011       Manual    322
21  2010       Manual    256
32  2016    Automatic    228
28  2014    Automatic    224
30  2015    Automatic    221
34  2017    Automatic    198
37  2018       Manual    191
26  2013    Automatic    179
24  2012    Automatic    142
22  2011    Automatic    130
17  2008       Manual    124
19  2009       Manual    124
36  2018    Automatic    107
15  2007       Manual     87
20  2010    Automatic     77
18  2009    Automatic     67
39  2019       Manual     56
13  2006       Manual     56
38  2019    Automatic     46
16  2008    Automatic     38
11  2005       Manual     33
9   2004       Manual     19
14  2007    Automatic     18
7   2003       Manual     13
5   2002       Manual     11
12  2006    Automatic     10
10  2005    Automatic      9
8   2004    Automatic      6
4   2001       Manual      4
3   2000       Manual      4
2   1999       Manual      2
1   1998       Manual      2
6   2003    Automatic      1
0   1998    Automatic      1

In [42]:
px.histogram(df.groupby(["Year","Transmission"])["count"].sum().reset_index().sort_values(by="count",ascending=False),x="Transmission",y="count",color="Year",text_auto=True,title="The Transmission of car has changed over the years")

In [43]:
df.Transmission.value_counts()

Transmission
Manual       4170
Automatic    1702
Name: count, dtype: int64

In [44]:
df.columns

Index(['Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type',
       'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats',
       'Price', 'Brand', 'count'],
      dtype='object')

________________________

# 4. Data understanding 

### 4.1 Univariate analysis 

In [45]:
px.histogram(df['Kilometers_Driven'])

In [46]:
px.histogram(df['Engine'])

In [47]:
px.histogram(df['Power'])

In [48]:
px.histogram(df['Mileage'])

In [49]:
px.histogram(df['Price'])

# Remove Outliers to check data distribution 

In [50]:
df_visual = df.copy()

In [51]:
numerical_cols = ['Kilometers_Driven' , 'Mileage' , 'Engine' , 'Power','Price']

In [52]:
from datasist.structdata import detect_outliers 

outliears = detect_outliers(df[numerical_cols],0,df[numerical_cols].columns)
df[numerical_cols].drop(outliears,inplace=True)

In [53]:
px.histogram(df_visual['Power'])

In [54]:
px.histogram(df_visual['Kilometers_Driven'])

In [55]:
px.histogram(df_visual['Mileage'])

In [56]:
px.histogram(df_visual['Engine'])

In [57]:
px.histogram(df, x="Location")

In [58]:
px.histogram(df, x="Year")

In [59]:
px.histogram(df, x="Fuel_Type")

In [60]:
px.histogram(df, x="Transmission")

In [61]:
px.histogram(df, x="Owner_Type")

In [62]:
px.histogram(df, x="Brand")

### 4.2 Univariate analysis 

- year
- kilometers 
- transmission 
- brand 

## Data summery : 
    - Numerical columns follow the gussian distribution, but have Outliers 
    (My desicion is not to remove it,because it will ne useful)
    - The countrt with the most sales spread is `Mumbai` and the least is `Ahmedabad`
    - Most of cars is form 2010 and 2015

In [63]:
df.head()

Name    Location  Year  Kilometers_Driven Fuel_Type Transmission  \
0   Maruti Wagon      Mumbai  2010              72000       CNG       Manual   
1  Hyundai Creta        Pune  2015              41000    Diesel       Manual   
2     Honda Jazz     Chennai  2011              46000    Petrol       Manual   
3  Maruti Ertiga     Chennai  2012              87000    Diesel       Manual   
4        Audi A4  Coimbatore  2013              40670    Diesel    Automatic   

  Owner_Type  Mileage  Engine   Power  Seats  Price    Brand  count  
0      First    26.60   998.0   58.16    5.0   1.75   Maruti      1  
1      First    19.67  1582.0  126.20    5.0  12.50  Hyundai      1  
2      First    18.20  1199.0   88.70    5.0   4.50    Honda      1  
3      First    20.77  1248.0   88.76    7.0   6.00   Maruti      1  
4     Second    15.20  1968.0  140.80    5.0  17.74     Audi      1

________________________

# 5. Data pre-processing

### 5.1 Data splitting 

In [64]:
# define dataset
X, y = df.drop("Price",axis=1) , df["Price"] 

# split into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=1)

In [65]:
X_train

Name    Location  Year  Kilometers_Driven Fuel_Type  \
5975      Hyundai i20  Coimbatore  2017              14618    Petrol   
4117          Audi Q5   Bangalore  2015              63000    Diesel   
1727   Chevrolet Aveo        Pune  2007              59000    Petrol   
4310       Tata Manza      Jaipur  2011              86500    Diesel   
2483    Toyota Innova       Delhi  2011              72000    Diesel   
...               ...         ...   ...                ...       ...   
935    Renault Duster      Jaipur  2015              91000    Diesel   
5325     Maruti Swift       Kochi  2016              65210    Diesel   
4088    Skoda Octavia   Hyderabad  2015             122086    Diesel   
244     Toyota Innova      Mumbai  2014             118000    Diesel   
5290  Volkswagen Polo   Hyderabad  2016              29647    Diesel   

     Transmission Owner_Type  Mileage  Engine   Power  Seats       Brand  \
5975       Manual      First    18.60  1197.0   81.83    5.0     Hyundai   
4117    Automatic      First    14.16  1968.0  174.30    5.0        Audi   
1727       Manual      Third    15.26  1150.0   74.90    5.0   Chevrolet   
4310       Manual     Second    21.12  1248.0   88.80    5.0        Tata   
2483       Manual      First    12.80  2494.0  102.00    7.0      Toyota   
...           ...        ...      ...     ...     ...    ...         ...   
935        Manual      First    19.87  1461.0   83.80    5.0     Renault   
5325       Manual      First    22.90  1248.0   74.00    5.0      Maruti   
4088    Automatic      First    19.30  1968.0  141.00    5.0       Skoda   
244        Manual      First    12.80  2494.0  102.00    7.0      Toyota   
5290       Manual      First    20.14  1498.0   88.00    5.0  Volkswagen   

      count  
5975      1  
4117      1  
1727      1  
4310      1  
2483      1  
...     ...  
935       1  
5325      1  
4088      1  
244       1  
5290      1  

[4404 rows x 13 columns]

In [66]:
X_test

Name    Location  Year  Kilometers_Driven Fuel_Type  \
5495          Hyundai Creta      Mumbai  2015              26000    Diesel   
2228          Maruti Ertiga      Mumbai  2016              65263       CNG   
2150        Mahindra XUV500  Coimbatore  2016              67472    Diesel   
664           Hyundai Verna      Jaipur  2008             170000    Diesel   
570              Honda City     Kolkata  2011              42086    Petrol   
...                     ...         ...   ...                ...       ...   
1799  Mercedes-Benz E-Class   Bangalore  2013              29000    Diesel   
4154           Maruti Swift   Bangalore  2012              62000    Diesel   
725              Skoda Yeti      Jaipur  2016              84000    Diesel   
850           Toyota Innova        Pune  2017              43200    Diesel   
3262           Maruti Swift      Jaipur  2014              75074    Diesel   

     Transmission Owner_Type  Mileage  Engine   Power  Seats          Brand  \
5495    Automatic      First    17.01  1582.0  126.20    5.0        Hyundai   
2228       Manual      First    17.50  1373.0   80.46    7.0         Maruti   
2150       Manual      First    16.00  2179.0  140.00    7.0       Mahindra   
664        Manual     Second    16.80  1493.0  110.00    5.0        Hyundai   
570        Manual      First    17.00  1497.0  118.00    5.0          Honda   
...           ...        ...      ...     ...     ...    ...            ...   
1799    Automatic      First    14.21  2143.0  203.00    5.0  Mercedes-Benz   
4154       Manual      First    17.80  1248.0   75.00    5.0         Maruti   
725        Manual      First    17.72  1968.0  108.50    5.0          Skoda   
850        Manual      First    13.68  2393.0  147.80    7.0         Toyota   
3262       Manual      First    23.40  1248.0   74.00    5.0         Maruti   

      count  
5495      1  
2228      1  
2150      1  
664       1  
570       1  
...     ...  
1799      1  
4154      1  
725       1  
850       1  
3262      1  

[1468 rows x 13 columns]

> Why I have started with splliting data  ?
- That's to avoid data leakage problem. 

### 5.2 Data transformation 

#### 5.2.1 Numerical transformation 

In [67]:
from sklearn.preprocessing import RobustScaler

In [68]:
Numerical_data = ['Year','Kilometers_Driven','Mileage','Engine','Power','Seats']

In [69]:
# define the scaler
scaler = RobustScaler()
# fit on the training dataset
scaler.fit(X_train[Numerical_data])
# scale the training dataset
X_train[Numerical_data] = scaler.transform(X_train[Numerical_data])
# scale the test dataset
X_test[Numerical_data] = scaler.transform(X_test[Numerical_data])

In [70]:
X_train

Name    Location  Year  Kilometers_Driven Fuel_Type  \
5975      Hyundai i20  Coimbatore  0.75          -0.982777    Petrol   
4117          Audi Q5   Bangalore  0.25           0.274062    Diesel   
1727   Chevrolet Aveo        Pune -1.75           0.170152    Petrol   
4310       Tata Manza      Jaipur -0.75           0.884530    Diesel   
2483    Toyota Innova       Delhi -0.75           0.507858    Diesel   
...               ...         ...   ...                ...       ...   
935    Renault Duster      Jaipur  0.25           1.001429    Diesel   
5325     Maruti Swift       Kochi  0.50           0.331472    Diesel   
4088    Skoda Octavia   Hyderabad  0.25           1.808962    Diesel   
244     Toyota Innova      Mumbai  0.00           1.702819    Diesel   
5290  Volkswagen Polo   Hyderabad  0.50          -0.592363    Diesel   

     Transmission Owner_Type   Mileage    Engine     Power  Seats       Brand  \
5975       Manual      First  0.068847 -0.373266 -0.192868    0.0     Hyundai   
4117    Automatic      First -0.695353  0.598991  1.272583    0.0        Audi   
1727       Manual      Third -0.506024 -0.432535 -0.302694    0.0   Chevrolet   
4310       Manual     Second  0.502582 -0.308953 -0.082409    0.0        Tata   
2483       Manual      First -0.929432  1.262295  0.126783    2.0      Toyota   
...           ...        ...       ...       ...       ...    ...         ...   
935        Manual      First  0.287435 -0.040353 -0.161648    0.0     Renault   
5325       Manual      First  0.808950 -0.308953 -0.316957    0.0      Maruti   
4088    Automatic      First  0.189329  0.598991  0.744849    0.0       Skoda   
244        Manual      First -0.929432  1.262295  0.126783    2.0      Toyota   
5290       Manual      First  0.333907  0.006305 -0.095087    0.0  Volkswagen   

      count  
5975      1  
4117      1  
1727      1  
4310      1  
2483      1  
...     ...  
935       1  
5325      1  
4088      1  
244       1  
5290      1  

[4404 rows x 13 columns]

In [71]:
X_test

Name    Location  Year  Kilometers_Driven Fuel_Type  \
5495          Hyundai Creta      Mumbai  0.25          -0.687102    Diesel   
2228          Maruti Ertiga      Mumbai  0.50           0.332848       CNG   
2150        Mahindra XUV500  Coimbatore  0.50           0.390232    Diesel   
664           Hyundai Verna      Jaipur -1.50           3.053643    Diesel   
570              Honda City     Kolkata -0.75          -0.269230    Petrol   
...                     ...         ...   ...                ...       ...   
1799  Mercedes-Benz E-Class   Bangalore -0.25          -0.609170    Diesel   
4154           Maruti Swift   Bangalore -0.50           0.248084    Diesel   
725              Skoda Yeti      Jaipur  0.50           0.819587    Diesel   
850           Toyota Innova        Pune  0.75          -0.240291    Diesel   
3262           Maruti Swift      Jaipur  0.00           0.587713    Diesel   

     Transmission Owner_Type   Mileage    Engine     Power  Seats  \
5495    Automatic      First -0.204819  0.112232  0.510301    0.0   
2228       Manual      First -0.120482 -0.151324 -0.214580    2.0   
2150       Manual      First -0.378657  0.865069  0.729002    2.0   
664        Manual     Second -0.240964  0.000000  0.253566    0.0   
570        Manual      First -0.206540  0.005044  0.380349    0.0   
...           ...        ...       ...       ...       ...    ...   
1799    Automatic      First -0.686747  0.819672  1.727417    0.0   
4154       Manual      First -0.068847 -0.308953 -0.301109    0.0   
725        Manual      First -0.082616  0.598991  0.229794    0.0   
850        Manual      First -0.777969  1.134931  0.852615    2.0   
3262       Manual      First  0.895009 -0.308953 -0.316957    0.0   

              Brand  count  
5495        Hyundai      1  
2228         Maruti      1  
2150       Mahindra      1  
664         Hyundai      1  
570           Honda      1  
...             ...    ...  
1799  Mercedes-Benz      1  
4154         Maruti      1  
725           Skoda      1  
850          Toyota      1  
3262         Maruti      1  

[1468 rows x 13 columns]

#### 5.2.1 Catogerical transformation

In [72]:
Nominal_data = ['Name','Location','Fuel_Type','Transmission','Power','Brand']
ordinal_data = ['Owner_Type']

##### 5.2.1.1 ordinal transformation

> The best case of owner type is (First), Of course the worst is Fourth & Above

In [73]:
df['Owner_Type'].unique()

array(['First', 'Second', 'Fourth & Above', 'Third'], dtype=object)

In [74]:
transformation = {
    "First":3,
    "Second":2,
    "Third":1,
    "Fourth & Above":0
}

In [75]:
X_train['Owner_Type'] = X_train['Owner_Type'].map(transformation)

In [76]:
X_test['Owner_Type'] = X_test['Owner_Type'].map(transformation)

In [77]:
X_train

Name    Location  Year  Kilometers_Driven Fuel_Type  \
5975      Hyundai i20  Coimbatore  0.75          -0.982777    Petrol   
4117          Audi Q5   Bangalore  0.25           0.274062    Diesel   
1727   Chevrolet Aveo        Pune -1.75           0.170152    Petrol   
4310       Tata Manza      Jaipur -0.75           0.884530    Diesel   
2483    Toyota Innova       Delhi -0.75           0.507858    Diesel   
...               ...         ...   ...                ...       ...   
935    Renault Duster      Jaipur  0.25           1.001429    Diesel   
5325     Maruti Swift       Kochi  0.50           0.331472    Diesel   
4088    Skoda Octavia   Hyderabad  0.25           1.808962    Diesel   
244     Toyota Innova      Mumbai  0.00           1.702819    Diesel   
5290  Volkswagen Polo   Hyderabad  0.50          -0.592363    Diesel   

     Transmission  Owner_Type   Mileage    Engine     Power  Seats  \
5975       Manual           3  0.068847 -0.373266 -0.192868    0.0   
4117    Automatic           3 -0.695353  0.598991  1.272583    0.0   
1727       Manual           1 -0.506024 -0.432535 -0.302694    0.0   
4310       Manual           2  0.502582 -0.308953 -0.082409    0.0   
2483       Manual           3 -0.929432  1.262295  0.126783    2.0   
...           ...         ...       ...       ...       ...    ...   
935        Manual           3  0.287435 -0.040353 -0.161648    0.0   
5325       Manual           3  0.808950 -0.308953 -0.316957    0.0   
4088    Automatic           3  0.189329  0.598991  0.744849    0.0   
244        Manual           3 -0.929432  1.262295  0.126783    2.0   
5290       Manual           3  0.333907  0.006305 -0.095087    0.0   

           Brand  count  
5975     Hyundai      1  
4117        Audi      1  
1727   Chevrolet      1  
4310        Tata      1  
2483      Toyota      1  
...          ...    ...  
935      Renault      1  
5325      Maruti      1  
4088       Skoda      1  
244       Toyota      1  
5290  Volkswagen      1  

[4404 rows x 13 columns]

In [78]:
X_test

Name    Location  Year  Kilometers_Driven Fuel_Type  \
5495          Hyundai Creta      Mumbai  0.25          -0.687102    Diesel   
2228          Maruti Ertiga      Mumbai  0.50           0.332848       CNG   
2150        Mahindra XUV500  Coimbatore  0.50           0.390232    Diesel   
664           Hyundai Verna      Jaipur -1.50           3.053643    Diesel   
570              Honda City     Kolkata -0.75          -0.269230    Petrol   
...                     ...         ...   ...                ...       ...   
1799  Mercedes-Benz E-Class   Bangalore -0.25          -0.609170    Diesel   
4154           Maruti Swift   Bangalore -0.50           0.248084    Diesel   
725              Skoda Yeti      Jaipur  0.50           0.819587    Diesel   
850           Toyota Innova        Pune  0.75          -0.240291    Diesel   
3262           Maruti Swift      Jaipur  0.00           0.587713    Diesel   

     Transmission  Owner_Type   Mileage    Engine     Power  Seats  \
5495    Automatic           3 -0.204819  0.112232  0.510301    0.0   
2228       Manual           3 -0.120482 -0.151324 -0.214580    2.0   
2150       Manual           3 -0.378657  0.865069  0.729002    2.0   
664        Manual           2 -0.240964  0.000000  0.253566    0.0   
570        Manual           3 -0.206540  0.005044  0.380349    0.0   
...           ...         ...       ...       ...       ...    ...   
1799    Automatic           3 -0.686747  0.819672  1.727417    0.0   
4154       Manual           3 -0.068847 -0.308953 -0.301109    0.0   
725        Manual           3 -0.082616  0.598991  0.229794    0.0   
850        Manual           3 -0.777969  1.134931  0.852615    2.0   
3262       Manual           3  0.895009 -0.308953 -0.316957    0.0   

              Brand  count  
5495        Hyundai      1  
2228         Maruti      1  
2150       Mahindra      1  
664         Hyundai      1  
570           Honda      1  
...             ...    ...  
1799  Mercedes-Benz      1  
4154         Maruti      1  
725           Skoda      1  
850          Toyota      1  
3262         Maruti      1  

[1468 rows x 13 columns]

##### 5.2.1.2 Nominal transformation


In [79]:
import category_encoders as ce
binaryencoder = ce.BinaryEncoder(cols=Nominal_data)
binaryencoder.fit(X_train)

BinaryEncoder(cols=['Name', 'Location', 'Fuel_Type', 'Transmission', 'Power',
                    'Brand'],
              mapping=[{'col': 'Name',
                        'mapping':       Name_0  Name_1  Name_2  Name_3  Name_4  Name_5  Name_6  Name_7
 1         0       0       0       0       0       0       0       1
 2         0       0       0       0       0       0       1       0
 3         0       0       0       0       0       0       1       1
 4         0       0       0       0       0       1       0       0
 5         0       0       0       0       0       1       0       1
...      ...     ...     ...     ...     ...     ...     ...     ...
 200       1       1       0       0       1       0       0       0
 201       1       1       0       0       1       0       0       1
 202       1       1       0       0       1       0       1       0
-1         0       0       0       0       0       0       0       0
-2         0       0       0       0       0       0       0       0

[204 rows x 8 columns]},
                       {'col': 'Location',
                        'mapping':      Location_0  Loca...
-1          0        0        0        0        0        0        0        0   
-2          0        0        0        0        0        0        0        0   

      Power_8  
 1          1  
 2          0  
 3          1  
 4          0  
 5          1  
...       ...  
 346        0  
 347        1  
 348        0  
-1          0  
-2          0  

[350 rows x 9 columns]},
                       {'col': 'Brand',
                        'mapping':      Brand_0  Brand_1  Brand_2  Brand_3  Brand_4
 1         0        0        0        0        1
 2         0        0        0        1        0
 3         0        0        0        1        1
 4         0        0        1        0        0
 5         0        0        1        0        1
 6         0        0        1        1        0
 7         0        0        1        1        1
 8         0        1        0        0        0
 9         0        1        0        0        1
 10        0        1        0        1        0
 11        0        1        0        1        1
 12        0        1        1        0        0
 13        0        1        1        0        1
 14        0        1        1        1        0
 15        0        1        1        1        1
 16        1        0        0        0        0
 17        1        0        0        0        1
 18        1        0        0        1        0
 19        1        0        0        1        1
 20        1        0        1        0        0
 21        1        0        1        0        1
 22        1        0        1        1        0
 23        1        0        1        1        1
 24        1        1        0        0        0
 25        1        1        0        0        1
 26        1        1        0        1        0
 27        1        1        0        1        1
 28        1        1        1        0        0
 29        1        1        1        0        1
-1         0        0        0        0        0
-2         0        0        0        0        0}])

In [80]:
X_test = binaryencoder.transform(X_test)

In [81]:
X_train = binaryencoder.transform(X_train)

In [82]:
X_test.columns

Index(['Name_0', 'Name_1', 'Name_2', 'Name_3', 'Name_4', 'Name_5', 'Name_6',
       'Name_7', 'Location_0', 'Location_1', 'Location_2', 'Location_3',
       'Year', 'Kilometers_Driven', 'Fuel_Type_0', 'Fuel_Type_1',
       'Fuel_Type_2', 'Transmission_0', 'Transmission_1', 'Owner_Type',
       'Mileage', 'Engine', 'Power_0', 'Power_1', 'Power_2', 'Power_3',
       'Power_4', 'Power_5', 'Power_6', 'Power_7', 'Power_8', 'Seats',
       'Brand_0', 'Brand_1', 'Brand_2', 'Brand_3', 'Brand_4', 'count'],
      dtype='object')

In [83]:
X_train

Name_0  Name_1  Name_2  Name_3  Name_4  Name_5  Name_6  Name_7  \
5975       0       0       0       0       0       0       0       1   
4117       0       0       0       0       0       0       1       0   
1727       0       0       0       0       0       0       1       1   
4310       0       0       0       0       0       1       0       0   
2483       0       0       0       0       0       1       0       1   
...      ...     ...     ...     ...     ...     ...     ...     ...   
935        0       0       1       0       1       0       0       1   
5325       0       0       1       0       0       0       0       1   
4088       0       0       0       0       1       1       0       0   
244        0       0       0       0       0       1       0       1   
5290       0       0       0       0       1       1       1       0   

      Location_0  Location_1  ...  Power_6  Power_7  Power_8  Seats  Brand_0  \
5975           0           0  ...        0        0        1    0.0        0   
4117           0           0  ...        0        1        0    0.0        0   
1727           0           0  ...        0        1        1    0.0        0   
4310           0           1  ...        1        0        0    0.0        0   
2483           0           1  ...        1        0        1    2.0        0   
...          ...         ...  ...      ...      ...      ...    ...      ...   
935            0           1  ...        1        1        0    0.0        1   
5325           1           0  ...        1        1        1    0.0        0   
4088           0           1  ...        1        0        1    0.0        0   
244            1           0  ...        1        0        1    2.0        0   
5290           0           1  ...        0        0        0    0.0        0   

      Brand_1  Brand_2  Brand_3  Brand_4  count  
5975        0        0        0        1      1  
4117        0        0        1        0      1  
1727        0        0        1        1      1  
4310        0        1        0        0      1  
2483        0        1        0        1      1  
...       ...      ...      ...      ...    ...  
935         0        0        0        1      1  
5325        0        1        1        0      1  
4088        1        0        0        1      1  
244         0        1        0        1      1  
5290        1        0        1        1      1  

[4404 rows x 38 columns]

# 6.0 Modeling 

### 6.1 Linear regression

In [84]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression().fit(X_train,np.log(y_train))
reg.score(X_test,  np.log(y_test))

0.8480038148379908

In [85]:
reg.score(X_train,np.log(y_train))

0.8599565234900445